packager/packager.cc

```diff
@@ -1542,6 +1556,83 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
+    UnorderedMap<core::packages::MangledName, bool> onStack;
+    std::vector<core::packages::MangledName> stack;
+};
+
+// DFS traversal for Tarjan's algorithm starting from pkgName, along with keeping track of some metadata needed for
+// detecting SCCs.
+void strongConnect(core::GlobalState &gs, ComputeSCCsMetadata &metadata, core::packages::MangledName pkgName) {
+    if (!gs.packageDB().getPackageInfo(pkgName).exists()) {
+        return;
+    }
+    auto &pkgInfo = PackageInfoImpl::from(gs.packageDB().getPackageInfo(pkgName));
+    metadata.index[pkgName] = metadata.nextIndex;
+    metadata.lowLink[pkgName] = metadata.nextIndex;
+    metadata.nextIndex++;
+    metadata.stack.push_back(pkgName);
+    metadata.onStack[pkgName] = true;
+    for (auto &i : pkgInfo.importedPackageNames) {
+        if (i.type == ImportType::Test) {
+            continue;
+        }
+        if (metadata.index[i.name.mangledName] == 0) {
+            // This is a tree edge (ie. a forward edge that we haven't visited yet).
+            strongConnect(gs, metadata, i.name.mangledName);
+            if (metadata.index[i.name.mangledName] == 0) {
+                // this might mean the other package doesn't exist, but that should have been caught already
+                continue;
+            }
+            // Since we can follow any number of tree edges for lowLink, the lowLink of child is valid for this package
+            // too.
+            metadata.lowLink[pkgName] = std::min(metadata.lowLink[pkgName], metadata.lowLink[i.name.mangledName]);
+        } else if (metadata.onStack[i.name.mangledName]) {
+            // This is a back edge (edge to ancestor) or cross edge (edge to a different subtree). Since we can only
+            // follow at most one back/cross edge, the best update we can make to lowlink of the current package is the
+            // child's index.
+            metadata.lowLink[pkgName] = std::min(metadata.lowLink[pkgName], metadata.index[i.name.mangledName]);
+        }
+        // If the child package is already visited and not on the stack, it's in a different SCC, so no update to the
+        // lowlink.
+    }
+
+    if (metadata.index[pkgName] == metadata.lowLink[pkgName]) {
+        // This is the root of an SCC. This means that all packages on the stack from this package to the top of the top
+        // of the stack are in the same SCC. Pop the stack until we reach the root of the SCC, and assign them the same
+        // SCC ID.
+        core::packages::MangledName poppedPkgName;
+        do {
+            poppedPkgName = metadata.stack.back();
+            metadata.stack.pop_back();
+            metadata.onStack[poppedPkgName] = false;
+            PackageInfoImpl &poppedPkgInfo = PackageInfoImpl::from(gs.packageDB().getPackageInfo(poppedPkgName));
+            poppedPkgInfo.sccID = metadata.nextSCCId;
+        } while (poppedPkgName != pkgName);
+        metadata.nextSCCId++;
```

Comment:
You could remember the size of the stack when you enter the function, and then iterate that range directly without popping in the loop. Then you could use one call to the `erase` method to drop all the names at once, instead of incrementally resizing it down.

Reply:
Does that make a difference? I don't think that `pop_back` changes the capacity of the vector, which would seem to me like it's just decrementing the `length` field in the vector.